thetajwar2003/learning-made-simple-ui
# Assignment 6 Submission: Nafis Khan

### Test Coverage via Jest:
<p align="center">
<img width='500' src='https://github.com/thetajwar2003/learning-made-simple-ui/assets/49691160/aec27e20-5253-4861-b60e-4dc4bb298108'>
</p

</br>

### Report Breakdown:
- Focus for Tests: Tests are focused on mainly the asks from the assignment therefore not all the components are covered at this moment.
- All files: Across all files, 9.43% of statements, 0% of branches, 18.18% of functions, and 9.61% of lines are covered by tests.
- src/app:
  - page.tsx: Navigation to Auth Page Tested.
- src/app/authentication/page.tsx: Login Form, Signup Form and and DB Tested.
  - GuestHeader.tsx: Navigation after authorization and log out Tested.

---

### Details of Capabilities Tested:
#### Unit Tests:
- src/app/page.tsx:
  - Tested the rendering of the main page component.
  - Ensured that specific elements are present on the page when it loads.
  - Checked the functionality of any interactive elements on the page.
- src/components/GuestHeader.tsx:
  - Verified that the guest header renders correctly.
  - Checked that all expected links and navigation elements are present.

#### Integration Tests:
- src/app/authentication/page.tsx:
  - Conducted tests to ensure that the authentication flow works as expected.
  - Checked the integration with any external authentication services.
  - Validated the handling of authentication errors and success states.
- src/app/dashboard/page.tsx:
  - Tested the integration of the dashboard page with backend services.
  - Ensured that data loads correctly and is displayed to the user.
  - Verified the functionality of any interactive elements on the dashboard.

#### End-to-End Tests:
- User Navigation and Authentication:
  - Home Page: Automated tests to open and verify the contents of the home page.
  - Login Page Navigation: Tested navigation from the home page to the login page, ensuring the transition is smooth and elements are loaded correctly.
  - User Login: Performed a user login with valid credentials and ensured the authentication process completes successfully.
  - Dashboard Redirection: After a successful login, verified that the user is correctly redirected to the dashboard page and that the dashboard content loads as expected.
 
#### Github Action Workflow:
- All Tests currently run using a GitHub Action on PRs made. Since the code for the workflow isn't merged to this repo, an example of the workflow can be seen in a [PR](https://github.com/nafisk/learning-made-simple-ui/pull/5) made to the forked repo. 
<p align="center">
<img width="500" alt="image" src="https://github.com/thetajwar2003/learning-made-simple-ui/assets/49691160/2928c0d3-4d79-4486-bc63-462a9ea54ab0">
</p

</br>

---


### Summary:
The overall code coverage is low, due to just starting the project. The main idea for tests is to make sure the important parts of the application work as expected. That is what I tried to do. I will aim for higher code coverage as the app becomes larger to ensure that your codebase is well-tested, which can lead to fewer bugs and more stable software.

---


### Test Project Locally:

1. **Please Clone forked Repository for the best experience:**
   - Run `git clone https://github.com/nafisk/learning-made-simple-ui` to clone the project repository to your local machine.
2. **Add .env File:**
   - Download the .env file from [this link](https://drive.google.com/file/d/1LHnqm4-abwlXju7XsQZ6phuo1tRDBBHB/view?usp=sharing).
   - Move the downloaded .env file to the project's root directory.
   - (Note: The .env file will be removed after grading.)
3. **Install Dependencies:**
   - Run `npm install` to install all the necessary dependencies for the project.
4. **Start the Server:**
   - Run `npm run dev` to start the server. (Ensure the server is running before proceeding to tests.)
5. **Run Tests:**
   - Use the following npm scripts to execute tests:
     - `npm run test`: Run all tests.
     - `npm run test:unit`: Run only unit tests.
     - `npm run test:integration`: Run only integration tests.
     - `npm run cypress:headless`: Run only e2e tests.

Generating the code coverage report using:
- `npm run test:coverage`: This will run the tests and generate a coverage report.

By following these setup steps and running the tests, you should be able to validate the latest changes and ensure that the new tests are contributing to the improved code coverage of the application.
